feat: add oc-ownerid and oc-permissions headers on PUT DAV requests

salmart-dev opened the pull request:
<!--
  - 🚨 SECURITY INFO
  -
  - Before sending a pull request that fixes a security issue please report it via our HackerOne page (https://hackerone.com/nextcloud) following our security policy (https://nextcloud.com/security/). This allows us to coordinate the fix and release without potentially exposing all Nextcloud servers and users in the meantime.
-->

* Resolves: #54088

## Summary

Adds `OC-Permissions` and `OC-OwnerId` headers to the response of PUT DAV requests.

Documentation update: https://github.com/nextcloud/documentation/pull/13525

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

salmart-dev committed:
feat: add oc-ownerid and oc-permissions headers on PUT DAV requests

Signed-off-by: Salvatore Martire <4652631+salmart-dev@users.noreply.github.com>

come-nc reviewed (changes requested):
New non-standard headers should be prefixed with `X-`

salmart-dev commented on apps/dav/lib/Connector/Sabre/AddExtraHeadersPlugin.php:
```diff
@@ -0,0 +1,64 @@
+
+	private LoggerInterface $logger;
+
+	public function __construct(LoggerInterface $logger) {
```
Done

come-nc commented on apps/dav/lib/Connector/Sabre/AddExtraHeadersPlugin.php:
```diff
@@ -0,0 +1,64 @@
+
+		$ownerId = $node->getOwner()?->getUID();
+		if ($ownerId !== null) {
+			$response->setHeader('OC-OwnerId', $ownerId);
```
Ah, maybe I’m wrong that this is a new header, do we already use OC-OwnerId in other places?

come-nc commented on apps/dav/lib/Server.php:
```diff
@@ -382,6 +383,7 @@ public function __construct(
 						)
 					);
 				}
+				$this->server->addPlugin(new AddExtraHeadersPlugin($logger));
```
As for other sabre plugin related features, are you certain this is the only place that needs it?
What about ServerFactory, publicremote.php, publicwebdav.php?

salmart-dev replied:
Added in `ServerFactory`, publicremote and publicwebdav instantiate the server through it. Thanks for the second catch :upside_down_face:

salmart-dev commented on apps/dav/lib/Connector/Sabre/AddExtraHeadersPlugin.php:
```diff
@@ -0,0 +1,63 @@
+			$this->logger->error("Cannot set extra headers for node of type {$nodeType} for file '{$request->getPath()}'");
+			return;
+		}
+
```
Right, the ownerId is exposed in the `FilesPlugin`, same with the permissions.

For permissions there is a check for `isPublic` but in the two usages I found that is set to false, so I don't think that needs to be ported.

salmart-dev commented on apps/dav/tests/unit/Connector/Sabre/AddExtraHeadersPluginTest.php:
```diff
@@ -0,0 +1,134 @@
+<?php
+
+/**
```
Added

salmart-dev committed:
refactor: extract logic to hide mounts and share information

The permission string for directories and files can contain M or S
depending if they are respectively coming from a mount or a share. This
information is not to be disclosed when the share is a public one.

Signed-off-by: Salvatore Martire <4652631+salmart-dev@users.noreply.github.com>

salmart-dev commented:
The last three commits add the changes necessary to add the filtering logic to hide mount and share information from the permission string, under Robin's suggestion that the behaviour should be kept.

salmart-dev commented:
Last update :crossed_fingers: : the fixes with `$isPublic` having a wrong value have been merged and now they are usable for this PR. The Owner ID header is not set in the case `isPublic` is true.

provokateurin reviewed (approved):
I don't like traits :see_no_evil:

provokateurin reviewed (approved):
Nice :tada:

skjnldsv committed:
Merge branch 'master' into 54088-providePermissionsAndOwnerIdInPut

Signed-off-by: John Molakvoæ <skjnldsv@users.noreply.github.com>
